File: src/lib/keychain.ts
import { invoke } from "@tauri-apps/api/core";
import type { ApiKeyProvider } from "../types/user";

/** Saves an API key to the OS Keychain for the given provider. */
export async function setApiKey(
  provider: ApiKeyProvider,
  key: string
): Promise<void> {
  await invoke("set_api_key", { provider, key });
}

/**
 * Returns the stored API key for the given provider, or null if not set.
 * Keys are stored in the OS Keychain and never written to disk by this app.
 */
export async function getApiKey(
  provider: ApiKeyProvider
): Promise<string | null> {
  return await invoke<string | null>("get_api_key", { provider });
}

/** Removes the stored API key for the given provider. */
export async function deleteApiKey(provider: ApiKeyProvider): Promise<void> {
  await invoke("delete_api_key", { provider });
}

File: src-tauri/src/keychain.rs
use keyring::{Entry, Error as KeyringError};

const SERVICE: &str = "reading-buddy";

fn entry(provider: &str) -> Result<Entry, String> {
    Entry::new(SERVICE, provider).map_err(|e| e.to_string())
}

/// Saves an API key for the given provider to the OS Keychain.
#[tauri::command]
pub fn set_api_key(provider: String, key: String) -> Result<(), String> {
    entry(&provider)?
        .set_password(&key)
        .map_err(|e| e.to_string())
}

/// Returns the stored API key for the given provider, or None if not set.
#[tauri::command]
pub fn get_api_key(provider: String) -> Result<Option<String>, String> {
    match entry(&provider)?.get_password() {
        Ok(key) => Ok(Some(key)),
        Err(KeyringError::NoEntry) => Ok(None),
        Err(e) => Err(e.to_string()),
    }
}

/// Deletes the stored API key for the given provider.
/// Deleting a key that does not exist is a no-op.
#[tauri::command]
pub fn delete_api_key(provider: String) -> Result<(), String> {
    match entry(&provider)?.delete_password() {
        Ok(()) => Ok(()),
        Err(KeyringError::NoEntry) => Ok(()),
        Err(e) => Err(e.to_string()),
    }
}
